AirCasting/MicrophoneSession/MicrophoneManager.swift

```diff
@@ -0,0 +1,125 @@
+//
+
+import Foundation
+import AVFoundation
+import CoreAudio
+import CoreData
+
+class MicrophoneManager: ObservableObject {
+    var context: NSManagedObjectContext {
+        PersistenceController.shared.container.viewContext
+    }
+    var session: Session?
+    var measurementStream: MeasurementStream?
+    
+    private var recorder: AVAudioRecorder!
+    private var levelTimer = Timer()
+    
+    private var locationProvider = LocationProvider()
+    
+    func startRecording(session: Session) {
+        let audioSession = AVAudioSession.sharedInstance()
+        
+        //Check microphone permisson
+        if audioSession.recordPermission != .granted {
+            audioSession.requestRecordPermission { (isGranted) in
+                if !isGranted {
+                    // TODO: pop-up that informs user that we need access to mic with ability do go back to homescreen
+                    fatalError("You must allow audio recording for this demo to work")
+                }
+            }
+        }
+        
+        self.session = session
+        createDBStream()
+        locationProvider.requestLocation()
+        
+        let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
+        
+        let recordSettings: [String: Any] = [
+            AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
+            AVSampleRateKey: 44100.0,
+            AVNumberOfChannelsKey: 1,
+            AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
+        ]
+        
+        do {
+            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
+            try audioSession.setActive(true)
+            try recorder = AVAudioRecorder(url:url, settings: recordSettings)
+            
+        } catch {
+            return
+        }
+        
+        recorder.isMeteringEnabled = true
+        recorder.record()
+        
+        levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
+    }
+    
+    @objc func getMeasurement() {
+        recorder.updateMeters()
+        
+        let level = recorder.averagePower(forChannel: 0)
+        let startingLocation = obtainCurrentLocation()
+        
+        let measurement = Measurement(context: self.context)
+        
+        measurement.latitude = startingLocation.latitude
+        measurement.longitude = startingLocation.longitude
+        measurement.value = Double(level)
+        measurement.time = Date()
+        
+        session?.endTime = measurement.time
+        session?.status = 0
+        
+        measurementStream?.addToMeasurements(measurement)
+        
+        print(measurement)
+        
+        do {
+            try context.save()
+        } catch {
+            print(error)
+        }
+    }
+    
+    func stopRecording() {
+        levelTimer.invalidate()
+    }
+    
+    private
+    
+    func createDBStream() {
+        if let session = session {
```

Comment:
I'm not sure if that's the case in our app, depends if the mobile mic session should refresh recording automatically after f.e. force quitting the app. 
If yes, you do not check if this stream already exists in the database, you create a new one every time.

Reply:
We would need some kind of resumeRecording function for that and if that's the behavior we want we should implement it in a separate PR.